### PR TITLE
Fix extensions

### PR DESCRIPTION
**Pull request by tkruse**
@itamaro, @mattyclarkson, @massenz
Can someone review these changes?

Basically all this does is to allow different file extensions, similar to #17, but with more tests

### CODE REVIEW

File: cpplint.py [hunk 1]
```diff
@@ -6064,7 +6093,7 @@ def ProcessFileData(filename, file_extension, lines, error,
   CheckForIncludeWhatYouUse(filename, clean_lines, include_state, error)
 
   # Check that the .cc file has included its header if it exists.
```
Comment (itamaro) [inline cpplint.py]: change also the comment accordingly

File: cpplint.py [hunk 2]
```diff
@@ -52,10 +52,20 @@
 import sys
 import unicodedata
 
+# Files with any of these extensions are considered to be
+# header files (and will undergo different style checks).
+# This set can be extended by using the --headers
+# option (also supported in CPPLINT.cfg)
+_header_extensions = ['h', 'hpp', 'hxx', 'h++', 'cuh']
```
Comment (itamaro) [inline cpplint.py]: why did you choose to use lists over sets here?

File: cpplint.py [hunk 3]
```diff
@@ -4548,8 +4559,16 @@ def _DropCommonSuffixes(filename):
   Returns:
     The filename with the common suffix removed.
   """
-  for suffix in ('test.cc', 'regtest.cc', 'unittest.cc',
```
Comment (itamaro) [inline cpplint.py]: IIUC, you're rewriting this simple `for suffix` with `for bad_suffixes` that you fill up, because you need to generate it based on combinations of possible header/source extensions, and test file suffixes.

if this is the case, I think it would be more readable to just "say that":

    for suffix in itertools.chain(
        ('%s.%s' % (test_suffix.lstrip('_'), ext)
         for test_suffix, ext in itertools.product(_test_suffixes, _nonheader_extensions)),
        ('%s.%s' % (suffix, ext)
         for suffix, ext in itertools.product(['inl', 'imp', 'internal'], _header_extensions)))

File: cpplint.py [hunk 4]
```diff
@@ -4565,12 +4584,11 @@ def _IsTestFilename(filename):
   Returns:
     True if 'filename' looks like a test, False otherwise.
   """
-  if (filename.endswith('_test.cc') or
-      filename.endswith('_unittest.cc') or
-      filename.endswith('_regtest.cc')):
-    return True
-  else:
-    return False
+  for suffix in _test_suffixes:
```
Comment (itamaro) [inline cpplint.py]: another candidate for `itertools.product`

File: cpplint.py [hunk 5]
```diff
@@ -52,10 +52,20 @@
...
+
+
 # The allowed extensions for file names
 # This is set by --extensions flag.
-_valid_extensions = set(['c', 'cc', 'cpp', 'cxx', 'c++', 'h', 'hpp', 'hxx',
-    'h++'])
+_valid_extensions = set(_nonheader_extensions + _header_extensions)
+
+# files with this suffix before the extension will be treated as test files
+_test_suffixes = set(['_unittest', '_test', '_regtest'])
```
Comment (itamaro) [inline cpplint.py]: is this configurable using flags / CPPLINT.cfg?

Comment (tkruse) [inline cpplint.py]: so, that may be a nice additional feature, but maybe that would go well into #11

**Comment by tkruse** [comment]
Excellent review, I made the changes you suggested as additional commit, please look at it when you find the time

File: cpplint.py [hunk 6]
```diff
@@ -50,12 +50,23 @@
...
+# option (also supported in CPPLINT.cfg)
+_header_extensions = set(['h', 'hpp', 'hxx', 'h++', 'cuh'])
+_nonheader_extensions = set(['c', 'cc', 'cpp', 'cxx', 'c++', 'cu'])
+
+
 # The allowed extensions for file names
 # This is set by --extensions flag.
-_valid_extensions = set(['c', 'cc', 'cpp', 'cxx', 'c++', 'h', 'hpp', 'hxx',
-    'h++'])
+_valid_extensions = _nonheader_extensions.union(_header_extensions)
```
Comment (itamaro) [inline cpplint.py]: what's troubling me now is that this variable will be overridden by `--extensions` (or the CPPLINT.cfg `extensions=...`), breaking the relation between it and the `header_extensions` and `nonheader_extensions` variables. this will cause some unexpected behaviors...

I suggest that instead you enforce consistency, by not allowing this variable to be overridden directly, but by replacing it with flags and CPPLINT.cfg options for the separate sets (e.g. `--headers` and `--nonheaders`), and making sure to update `_valid_extensions` accordingly when modifying these.

File: cpplint.py [hunk 7]
```diff
@@ -50,12 +50,23 @@
 import sre_compile
 import string
 import sys
+import itertools
```
Comment (itamaro) [inline cpplint.py]: lint the linter ;-)

imports should be sorted

File: cpplint.py (hunk 6 above)
Comment (tkruse) [inline cpplint.py]: I think the confusing and unexpected behaviors already exist in the system as it is, because it uses header extensions throughout the code. So while you are right, this PR does not make matters worse than they are. So this would be for a separate PR, I think

File: cpplint.py (hunk 7 above)
Comment (tkruse) [inline cpplint.py]: sorted and squashed commits

File: cpplint.py (hunk 6 above)
Comment (itamaro) [inline cpplint.py]: I agree that the confusion already exists, but I disagree that this does not make matters worse.

By separating the extensions to several lists, and exposing only one of them to user-accessible configuration, making it possible that the lists go out of sync - I think this makes things worse (in some aspects) compared to the current state.

Unless I'm missing something...

I'll also be happy with leaving a TODO to fix this in a followup PR, assuming it's coming up soon :-)

Another option is to "unexpose" the `extensions` option until this gets settled in a followup PR.

Comment (tkruse) [inline cpplint.py]: fair enough. I want to stay backwards compatible, to changing the --extension option is not that great. I added a commit that circumvents the problem by not storing in global variables a mixture of defaults and user-specified files. Instead, at runtime functions evaluate those sets. This is more painful to review I guess, but you asked for it. Also this reminds me that I would like unit tests for all these file filters, but I won't have the time soon.

**Comment by tkruse** [comment]
now also added more text to the usage section as suggested in #17

**Comment by tkruse** [comment]
sorry, this does not work yet, I do need those tests

**Comment by tkruse** [comment]
@itamaro I wont have time to write tests this week. I fixed the problem though, and you may review the solution for the variables and the usage string.

**Comment by tkruse** [comment]
Superceded by https://github.com/theandrewdavis/cpplint/pull/14
